File: python/src/revenant/core/appearance/image.py
# pyright: reportUnknownMemberType=false
"""
Image loading and preparation for PDF signature fields.

Loads PNG/JPEG images, downscales if needed, extracts alpha channels,
and returns deflate-compressed pixel data ready for PDF embedding.
"""

from __future__ import annotations

import zlib
from pathlib import Path
from typing import TypedDict


class SignatureImageData(TypedDict):
    """Data returned by load_signature_image."""

    samples: bytes  # Deflate-compressed RGB pixel data
    smask: bytes | None  # Deflate-compressed alpha channel, or None if opaque
    width: int  # Pixel width
    height: int  # Pixel height
    bpc: int  # Bits per component (always 8)


# Maximum image dimension in pixels before downscaling.
# Signature images don't need to be large — 200x100 is plenty for print.
_MAX_IMAGE_PX = 200

# Maximum input file size (5 MB). Anything larger is certainly not a
# reasonable signature image and might be a decompression bomb or mistake.
_MAX_FILE_SIZE = 5 * 1024 * 1024

# Allowed image formats (Pillow format names).
_ALLOWED_FORMATS = {"PNG", "JPEG", "GIF", "BMP", "TIFF", "WEBP"}

# Maximum pixel count to prevent decompression bombs (CWE-400).
# Pillow's default (~89M pixels) allows ~270 MB RAM for RGB, which is
# excessive for signature images.  A 2000x2000 source is already far
# larger than the 200px target — anything beyond is suspicious.
_MAX_IMAGE_PIXELS = 2000 * 2000


def load_signature_image(image_path: str) -> SignatureImageData:
    """Load an image file and prepare it for embedding in a PDF signature.

    Supports PNG (with transparency) and JPEG.
    Images are downscaled if larger than 200px on any side.

    Args:
        image_path: Path to a PNG or JPEG file.

    Returns:
        dict with keys:
            'samples': bytes -- raw RGB pixel data (deflate-compressed)
            'smask': bytes or None -- alpha channel (deflate-compressed), None if opaque
            'width': int -- pixel width
            'height': int -- pixel height
            'bpc': int -- bits per component (always 8)

    Raises:
        FileNotFoundError: if image_path does not exist.
        ValueError: if the image format is not supported.
    """
    path = Path(image_path).resolve()
    if not path.exists():
        msg = f"Signature image not found: {path}"
        raise FileNotFoundError(msg)

    # Check file size before loading into memory
    file_size = path.stat().st_size
    if file_size > _MAX_FILE_SIZE:
        msg = (
            f"Signature image too large: {file_size / 1024 / 1024:.1f} MB "
            f"(max {_MAX_FILE_SIZE / 1024 / 1024:.0f} MB)"
        )
        raise ValueError(msg)
    if file_size == 0:
        raise ValueError("Signature image file is empty")

    try:
        from PIL import Image
    except ImportError as exc:
        raise ImportError(
            "Pillow is required for signature images.\n"
            "Install with: pip install pikepdf  (includes Pillow)"
        ) from exc

    try:
        img = Image.open(path)
    except OSError as exc:
        # PIL raises UnidentifiedImageError (subclass of OSError) for unknown formats,
        # and OSError for I/O issues.
        raise ValueError(f"Cannot load image file: {exc}") from exc

    try:
        # Reject images with excessive pixel count before any decompression.
        # Image.open() is lazy — it reads the header (dimensions) without
        # loading pixel data into memory.  Decompression happens during
        # resize/tobytes, so we must check BEFORE those calls.
        # A crafted file can be tiny on disk but decompress to massive
        # dimensions (decompression bomb, CWE-400).
        pixel_count = img.width * img.height
        if pixel_count > _MAX_IMAGE_PIXELS:
            raise ValueError(
                f"Image too large: {img.width}x{img.height} ({pixel_count:,} pixels). "
                f"Maximum: {_MAX_IMAGE_PIXELS:,} pixels."
            )

        # Validate image format — reject unknown formats (img.format is None
        # for raw streams) and explicitly unsupported ones.
        if not img.format or img.format not in _ALLOWED_FORMATS:
            actual = img.format or "unknown"
            msg = (
                f"Unsupported image format: {actual}. "
                f"Supported: {', '.join(sorted(_ALLOWED_FORMATS))}"
            )
            raise ValueError(msg)

        # Downscale if too large
        max_dim = max(img.width, img.height)
        if max_dim > _MAX_IMAGE_PX:
            scale = _MAX_IMAGE_PX / max_dim
            new_w = max(1, int(img.width * scale))
            new_h = max(1, int(img.height * scale))
            img = img.resize((new_w, new_h), Image.Resampling.LANCZOS)

        # Extract alpha channel if present
        smask_data = None
        if img.mode in ("RGBA", "LA", "PA"):
            alpha = img.split()[-1]
            smask_data = zlib.compress(alpha.tobytes())
            img = img.convert("RGB")
        elif img.mode != "RGB":
            img = img.convert("RGB")

        rgb_data = zlib.compress(img.tobytes())

        return {
            "samples": rgb_data,
            "smask": smask_data,
            "width": img.width,
            "height": img.height,
            "bpc": 8,
        }
    finally:
        img.close()

File: typescript/src/core/appearance/image.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Image loading and preparation for PDF signature fields.
 *
 * Loads PNG/JPEG images, downscales if needed, extracts alpha channels,
 * and returns deflate-compressed pixel data ready for PDF embedding.
 */

import { readFileSync, statSync } from "node:fs";
import { resolve } from "node:path";
import { deflateSync } from "node:zlib";

export interface SignatureImageData {
  samples: Uint8Array; // Deflate-compressed RGB pixel data
  smask: Uint8Array | null; // Deflate-compressed alpha channel
  width: number;
  height: number;
  bpc: number; // Bits per component (always 8)
}

/** Maximum image dimension in pixels before downscaling. */
const MAX_IMAGE_PX = 200;

/** Maximum input file size (5 MB). */
const MAX_FILE_SIZE = 5 * 1024 * 1024;

/** Maximum pixel count to prevent decompression bombs. */
const MAX_IMAGE_PIXELS = 2000 * 2000;

/**
 * Load an image file and prepare it for embedding in a PDF signature.
 *
 * Supports PNG (with transparency) and JPEG.
 * Images are downscaled if larger than 200px on any side.
 */
export async function loadSignatureImage(imagePath: string): Promise<SignatureImageData> {
  const fullPath = resolve(imagePath);

  let stat: ReturnType<typeof statSync>;
  try {
    stat = statSync(fullPath);
  } catch {
    throw new Error(`Signature image not found: ${fullPath}`);
  }

  if (stat.size > MAX_FILE_SIZE) {
    const sizeMb = (stat.size / 1024 / 1024).toFixed(1);
    const maxMb = (MAX_FILE_SIZE / 1024 / 1024).toFixed(0);
    throw new Error(`Signature image too large: ${sizeMb} MB (max ${maxMb} MB)`);
  }
  if (stat.size === 0) {
    throw new Error("Signature image file is empty");
  }

  const data = readFileSync(fullPath);

  // Detect format by magic bytes
  if (isPng(data)) {
    return loadPng(data);
  }
  if (isJpeg(data)) {
    return loadJpeg(data);
  }

  throw new Error("Unsupported image format. Supported: PNG, JPEG");
}

function isPng(data: Buffer): boolean {
  return (
    data.length > 8 && data[0] === 0x89 && data[1] === 0x50 && data[2] === 0x4e && data[3] === 0x47
  );
}

function isJpeg(data: Buffer): boolean {
  return data.length > 2 && data[0] === 0xff && data[1] === 0xd8;
}

async function loadPng(data: Buffer): Promise<SignatureImageData> {
  const { PNG } = await import("pngjs");

  const png = PNG.sync.read(data);

  if (png.width * png.height > MAX_IMAGE_PIXELS) {
    throw new Error(
      `Image too large: ${png.width}x${png.height}. Maximum: ${MAX_IMAGE_PIXELS} pixels.`,
    );
  }

  // Downscale if needed
  let pixels: Uint8Array | Buffer = png.data;
  let w = png.width;
  let h = png.height;
  const maxDim = Math.max(w, h);
  if (maxDim > MAX_IMAGE_PX) {
    const scale = MAX_IMAGE_PX / maxDim;
    const newW = Math.max(1, Math.round(w * scale));
    const newH = Math.max(1, Math.round(h * scale));
    pixels = nearestNeighborResize(pixels, w, h, newW, newH, 4);
    w = newW;
    h = newH;
  }

  // Extract alpha if present, separate RGB
  const pixelCount = w * h;
  const rgb = new Uint8Array(pixelCount * 3);
  let hasAlpha = false;
  let alpha: Uint8Array | null = null;

  // Check if any pixel has non-255 alpha
  for (let i = 0; i < pixelCount; i++) {
    if (pixels[i * 4 + 3] !== 255) {
      hasAlpha = true;
      break;
    }
  }

  if (hasAlpha) {
    alpha = new Uint8Array(pixelCount);
    for (let i = 0; i < pixelCount; i++) {
      const base = i * 4;
      rgb[i * 3] = pixels[base] ?? 0;
      rgb[i * 3 + 1] = pixels[base + 1] ?? 0;
      rgb[i * 3 + 2] = pixels[base + 2] ?? 0;
      alpha[i] = pixels[base + 3] ?? 0;
    }
  } else {
    for (let i = 0; i < pixelCount; i++) {
      const base = i * 4;
      rgb[i * 3] = pixels[base] ?? 0;
      rgb[i * 3 + 1] = pixels[base + 1] ?? 0;
      rgb[i * 3 + 2] = pixels[base + 2] ?? 0;
    }
  }

  return {
    samples: deflateSync(rgb),
    smask: alpha ? deflateSync(alpha) : null,
    width: w,
    height: h,
    bpc: 8,
  };
}

async function loadJpeg(data: Buffer): Promise<SignatureImageData> {
  const jpeg = await import("jpeg-js");
  const decoded = jpeg.decode(data, { useTArray: true });

  if (decoded.width * decoded.height > MAX_IMAGE_PIXELS) {
    throw new Error(
      `Image too large: ${decoded.width}x${decoded.height}. Maximum: ${MAX_IMAGE_PIXELS} pixels.`,
    );
  }

  let pixels: Uint8Array | Buffer = decoded.data;
  let w = decoded.width;
  let h = decoded.height;
  const maxDim = Math.max(w, h);
  if (maxDim > MAX_IMAGE_PX) {
    const scale = MAX_IMAGE_PX / maxDim;
    const newW = Math.max(1, Math.round(w * scale));
    const newH = Math.max(1, Math.round(h * scale));
    pixels = nearestNeighborResize(pixels, w, h, newW, newH, 4);
    w = newW;
    h = newH;
  }

  // JPEG has no alpha -- extract RGB only
  const pixelCount = w * h;
  const rgb = new Uint8Array(pixelCount * 3);
  for (let i = 0; i < pixelCount; i++) {
    const base = i * 4;
    rgb[i * 3] = pixels[base] ?? 0;
    rgb[i * 3 + 1] = pixels[base + 1] ?? 0;
    rgb[i * 3 + 2] = pixels[base + 2] ?? 0;
  }

  return {
    samples: deflateSync(rgb),
    smask: null,
    width: w,
    height: h,
    bpc: 8,
  };
}

/**
 * Nearest-neighbor resize for RGBA pixel data.
 */
function nearestNeighborResize(
  src: Uint8Array | Buffer,
  srcW: number,
  srcH: number,
  dstW: number,
  dstH: number,
  channels: number,
): Uint8Array {
  const dst = new Uint8Array(dstW * dstH * channels);
  for (let y = 0; y < dstH; y++) {
    const srcY = Math.floor((y * srcH) / dstH);
    for (let x = 0; x < dstW; x++) {
      const srcX = Math.floor((x * srcW) / dstW);
      const srcIdx = (srcY * srcW + srcX) * channels;
      const dstIdx = (y * dstW + x) * channels;
      for (let c = 0; c < channels; c++) {
        dst[dstIdx + c] = src[srcIdx + c] ?? 0;
      }
    }
  }
  return dst;
}
